runtime/src/accounts_db.rs

```diff
@@ -17709,6 +17709,123 @@ pub mod tests {
         }
     }
 
+    fn get_all_accounts(db: &AccountsDb, slots: Range<Slot>) -> Vec<(Pubkey, AccountSharedData)> {
+        slots
+            .clone()
+            .into_iter()
+            .filter_map(|slot| {
+                let storages = db.get_storages_for_slot(slot);
+                storages.map(|storages| {
+                    assert_eq!(storages.len(), 1, "slot: {}, slots: {:?}", slot, slots);
+                    let storage = storages.first().unwrap();
+                    storage
+                        .accounts
+                        .account_iter()
+                        .map(|account| (*account.pubkey(), account.to_account_shared_data()))
+                        .collect::<Vec<_>>()
+                })
+            })
+            .flatten()
+            .collect::<Vec<_>>()
+    }
+
+    fn compare_all_accounts(accounts: [&Vec<(Pubkey, AccountSharedData)>; 2]) {
+        let one = accounts[0];
+        let two = accounts[1];
+        let mut two_indexes = (0..two.len()).collect::<Vec<_>>();
+        one.iter().for_each(|(pubkey, account)| {
+            for i in 0..two_indexes.len() {
+                let pubkey2 = two[two_indexes[i]].0;
+                if pubkey2 == *pubkey {
+                    assert!(accounts_equal(account, &two[i].1));
+                    two_indexes.remove(i);
+                    break;
+                }
+            }
+        });
+        assert!(
+            two_indexes.is_empty(),
+            "one: {one:?}, two: {two:?}, two_indexes: {two_indexes:?}"
+        );
+    }
+
+    #[test]
+    fn test_shrink_ancient() {
+        solana_logger::setup();
+
+        let num_normal_slots = 1;
+        // build an ancient append vec at slot 'ancient_slot'
+        let (db, ancient_slot) = get_one_ancient_append_vec_and_others(true, num_normal_slots);
+
+        let max_slot_inclusive = ancient_slot + (num_normal_slots as Slot);
+        let initial_accounts = get_all_accounts(&db, ancient_slot..(max_slot_inclusive + 1));
+        compare_all_accounts([
+            &initial_accounts,
+            &get_all_accounts(&db, ancient_slot..(max_slot_inclusive + 1)),
+        ]);
+
+        // combine normal append vec(s) into existing ancient append vec
+        db.combine_ancient_slots(
+            (ancient_slot..=max_slot_inclusive).collect(),
+            CAN_RANDOMLY_SHRINK_FALSE,
+        );
+
+        compare_all_accounts([
+            &initial_accounts,
+            &get_all_accounts(&db, ancient_slot..(max_slot_inclusive + 1)),
```

Comment:
all the accounts should be in `ancient_slot` now, right? So should we get all accounts in `ancient_slot..=ancient_slot` instead?

Reply:
I want to scan ALL the slots. If there are duplicates or the accounts are in multiple places, I want to know that.

Reply:
That's a valid point. But I think we should also be checking that the accounts in normal slots actually got put into the ancient slot too. The way this test is currently written, it would pass even if `combine_ancient_slots` did nothing.

Probably best to compare both ranges then?